File: solutions/typescript/0097.ts
/**
 * 97. 交错字符串
 * 一维滚动 DP：dp[j] 表示 s3 的前 i+j 个字符能否由
 * s1 的前 i 个字符和 s2 的前 j 个字符交错组成（i 随外层循环推进）。
 * 长度不匹配直接返回 false。
 */
function isInterleave(s1: string, s2: string, s3: string): boolean {
    const m = s1.length;
    const n = s2.length;
    if (m + n !== s3.length) return false;

    const dp: boolean[] = new Array(n + 1).fill(false);
    dp[0] = true;
    for (let j = 1; j <= n; j++) {
        dp[j] = dp[j - 1] && s2[j - 1] === s3[j - 1];
    }

    for (let i = 1; i <= m; i++) {
        dp[0] = dp[0] && s1[i - 1] === s3[i - 1];
        for (let j = 1; j <= n; j++) {
            dp[j] = (dp[j] && s1[i - 1] === s3[i + j - 1]) || (dp[j - 1] && s2[j - 1] === s3[i + j - 1]);
        }
    }

    return dp[n];
}

export default isInterleave;

File: solutions/rust/0097.rs
impl Solution {
    /// 97. 交错字符串
    /// 一维滚动 DP：dp[j] 表示 s3 的前 i+j 个字符能否由
    /// s1 的前 i 个字符和 s2 的前 j 个字符交错组成（i 随外层循环推进）。
    /// 长度不匹配直接返回 false。
    pub fn is_interleave(s1: String, s2: String, s3: String) -> bool {
        let s1 = s1.as_bytes();
        let s2 = s2.as_bytes();
        let s3 = s3.as_bytes();
        let (m, n) = (s1.len(), s2.len());
        if m + n != s3.len() {
            return false;
        }

        let mut dp = vec![false; n + 1];
        dp[0] = true;
        for j in 1..=n {
            dp[j] = dp[j - 1] && s2[j - 1] == s3[j - 1];
        }

        for i in 1..=m {
            dp[0] = dp[0] && s1[i - 1] == s3[i - 1];
            for j in 1..=n {
                dp[j] = (dp[j] && s1[i - 1] == s3[i + j - 1]) || (dp[j - 1] && s2[j - 1] == s3[i + j - 1]);
            }
        }

        dp[n]
    }
}
